### PR TITLE
Rewrite the legacy C diptest code in C++

### PR DESCRIPTION
**Pull request by prokolyvakis**
## What?

1. The legacy C code that performs the Dip calculation has been rewritten from scratch in C++.
2. Extensive documentation was added to allow for the easier understanding of the algorithm's steps.

## Why?

The current code, written in C, that performs the dip calculation is (i) pretty difficult to understand and (ii) to maintain. I believe that the merge of this PR would allow even the easier addition of new features, i.e., safer data type, e.g., size_t.

## How?

- Both the source `diptest.c` and the header `diptest.h` files have been deleted.
- In their place a source file `diptest.cpp` and the relevant header `diptest.hpp` file have been added.
- The relevant changes in both the `CMakeList.txt` and the wrapper `wrapper.hpp` have been performed, i.e., include the newly created files instead of the old ones.

### Architectural Design Choices

After a first dip's algorithm examination, it became apparent that both the calculation of the lcm & gcm have a lot in common. For that reason, the `ConvexEnvelope` class has been created that contains all the needed variables to perform the lcm (or gcm) fit. To be on par with the execution time of the current C implementation various decisions have been taken which are shortly described in the following.

#### A. Only public class parameters and functions

To remove any overhead introduced by getters and setters and since the class itself is not exposed via the wrapper, all the class parameters and members have been defined as public. This is in accordance to the fact that the `diptst` subroutine provides direct references to all the parameters that the `ConvexEnvelope` class handles. Making the class's parameters private would require costly dynamic allocation and deep copies.

#### B. Loop unrolling

In certain cases, the speed was preferred rather than abstraction. One such example is the _collection of the change points for the GCM (or LCM) from `high` to `low`_ in the `diptest.cpp` (lines 315-317). Although that code segment could easily become a member function of the `ConvexEnvelope` class, I did not do it, since it would require either an if-else statement or a lambda function (for merging the statements: `gcm[i] > low` and `lcm[i] < high` into one). Both of them could be costly and of minor importance as far as the code understanding is concerned.

## Testing?

The implementation has been tested locally in a Darwin platform and all the unittests pass with success. I performed various speed checks and I did not notice any real difference.

## Final Words

Thank you for your great work in creating a nice Python lib for dip calculation!

### CODE REVIEW

**Comment by RUrlus** [comment]
Hi @prokolyvakis,

Thank you for the PR! I am definitively open to it, looks good.

It did notice there is a segfault and acces violation (Windows only), I am guessing that the we're returning something that is now being destructed. Regarding windows, I suspect that we have a different allocator/deallocator issue.

Could you rebase on the current stable? I just switch the RNG and hadn't yet noticed the PR, apologies.
I will have a proper look at the PR in a few days when I'm back from a business trip.

**Comment by prokolyvakis** [comment]
Hi @RUrlus,

I have just rebased on the current stable! Regarding the segfault (and access violation), I have to admit that did not use smart pointers or move semantics in the current implementation (neither in the subroutine nor in the newly created classes and functions), so there could be a deallocation bug. I did not manage to reproduce it locally, but I believe that I will easily solve it next week ( I will also be AFK up to the end of this week ).

In the meanwhile, I am looking forward to your code review and I wish a safe business trip!

File: include/diptest/diptest.hpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    {                 \
+        (void)(expr); \
+    } while (0)
+
+#include <iterator> // for iterators
+#include <vector>   // for vectors
+#include <assert.h> // for assert
+#include <iostream> // for cout
+#include <iomanip>  // for setw
+#include <math.h>   // for isgreaterequal
```
Comment (RUrlus) [inline include/diptest/diptest.hpp]: Prefer to use the C++ equivalents (cmath, cassert)

File: include/diptest/diptest.hpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    const ConvexEnvelopeType type;
+    
+    /// Constructors:
+
+    ConvexEnvelope(
+        const double *arr,
+        int *optimum,
+        int *indices,
+        int size,
+        ConvexEnvelopeType type) : arr(arr), optimum(optimum), indices(indices), size(size), type(type){};
```
Comment (RUrlus) [inline include/diptest/diptest.hpp]: Can you check the order, clangd gives me initialisation order warnings

Comment (RUrlus) [inline include/diptest/diptest.hpp]: Can you check the order, clangd gives me initialisation order warnings

File: src/diptest.cpp [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+ *
+ * @version 0.1
+ * @date 2022-04-19
+ *
+ */
+#include <diptest/diptest.hpp>
+
+void ConvexEnvelope::compute_indices()
+{
+    vector<int> range;
```
Comment (RUrlus) [inline src/diptest.cpp]: Why is the range being allocated? The C code doesn't allocate anything allowing us to reuse memory blocks in bootstrap computations. This adds overhead I don't think we need

Comment (RUrlus) [inline src/diptest.cpp]: Why is the range being allocated? The C code doesn't allocate anything allowing us to reuse memory blocks in bootstrap computations. This adds overhead I don't think we need

File: src/diptest.cpp [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    int sign = 1, offset = 0;
+    if (type == MAJORANT)
+    {
+        sign = -1;
+        offset = size + 1;
+    }
+
+    for (int i = 1; i <= size; i++)
+    {
+        range.push_back(offset + sign * i);
```
Comment (RUrlus) [inline src/diptest.cpp]: If we want/need to use the vector can we allocate it with the proper size, this is relatively slow

Comment (RUrlus) [inline src/diptest.cpp]: If we want/need to use the vector can we allocate it with the proper size, this is relatively slow

File: src/diptest.cpp [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+            if (opt_at_nxt_iter == range_start || rate_change_flag)
+                break;
+            indices[*next_idx] = opt_at_nxt_iter;
+        }
+    }
+}
+
+Dip ConvexEnvelope::compute_dip()
+{
+    int offset = (type == MINORANT) ? 0 : 1;
```
Comment (RUrlus) [inline src/diptest.cpp]: Can we make these const to help the compiler?

Comment (RUrlus) [inline src/diptest.cpp]: Can we make these const to help the compiler?

File: src/diptest.cpp [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+    int sign = 1 + -2 * offset;
```
Comment (RUrlus) [inline src/diptest.cpp]: Also const here

Comment (RUrlus) [inline src/diptest.cpp]: Also const here

File: src/diptest.cpp [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+        sign = -1;
+        offset = size + 1;
+    }
+
+    for (int i = 1; i <= size; i++)
+    {
+        range.push_back(offset + sign * i);
+    }
+
+    vector<int>::const_iterator curr_idx = range.begin();
```
Comment (RUrlus) [inline src/diptest.cpp]: I am wondering whether the overhead of this range based for-loop is worth it. We use plain for loops in other parts of the code.

Comment (RUrlus) [inline src/diptest.cpp]: I am wondering whether the overhead of this range based for-loop is worth it. We use plain for loops in other parts of the code.

**Review by RUrlus — COMMENTED** [review]
Hi @prokolyvakis,

Thanks again for the PR. I had a look at the code and added some questions remarks.
Besides the segfault, I noticed a performance regression. It's about 15% slower on my machine:
```python
import numpy as np
import diptest as dt
sample = np.random.normal(0, 1, 10000)
%timeit -r 100 dt.diptest(sample, boot_pval=True, n_boot=10000, n_threads=4)
```
yields
```
# with diptest.c (v0.3.0)
2.02 s ± 28.4 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)
# with diptest.cpp
2.23 s ± 56.4 ms per loop (mean ± std. dev. of 100 runs, 1 loop each) 
```
I am guessing this is mostly due to the range. For my application performance is quite important so I would to avoid such a drop.

File: include/diptest/diptest.hpp (hunk 1 above)
Comment (prokolyvakis) [reply include/diptest/diptest.hpp]: Nice catch! Resolved in the most recent pushed commit.

File: include/diptest/diptest.hpp (hunk 2 above)
Comment (prokolyvakis) [reply include/diptest/diptest.hpp]: I will look into that!

File: src/diptest.cpp (hunk 3 above)
Comment (prokolyvakis) [reply src/diptest.cpp]: Nice catch, in the current version all vector dependencies have been removed!

File: src/diptest.cpp (hunk 4 above)
Comment (prokolyvakis) [reply src/diptest.cpp]: Same comment as above!

Comment (prokolyvakis) [reply src/diptest.cpp]: Not needed anymore!

File: src/diptest.cpp (hunk 5 above)
Comment (prokolyvakis) [reply src/diptest.cpp]: Nice catch! Took into account!

File: src/diptest.cpp (hunk 6 above)
Comment (prokolyvakis) [reply src/diptest.cpp]: Same as above!

**Comment by prokolyvakis** [comment]
Hello @RUrlus 

Thanks for the code review. Regarding the segfault (and related issues), I think I resolved it. I tested it in multiple dockerized envs and in all the considered setups the test succeeded without any error any more. I would be grateful if you could rerun the tests to verify that this is the case.

Regarding your suggestions, I think I incorporated all of them (_again I would be grateful if you could double check if I forgot something_).

Finally, regarding the performance regression that you noticed I do not believe that the current version did anything on this regard. I do not believe that it's only related with the vector allocation (_as I said the current version removed all such dependencies so you can check if the performance drop was luckily resolved but from my experience we would need further optimization action items_)!

**Comment by RUrlus** [comment]
Hi @prokolyvakis,

Thanks for the fix!
Removing the vector actually had a pretty significant effect on the performance regression.
About half of the regression is gone now:

```
# with diptest.c (v0.3.0)
2.02 s ± 28.4 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)
# with diptest.cpp (old)
2.23 s ± 56.4 ms per loop (mean ± std. dev. of 100 runs, 1 loop each) 
# with diptest.cpp (260eb7)
2.1 s ± 21.1 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)
```
I had a look but don't see any obvious source for the regression at this point. I guess the zero-cost abstractions are not always zero-cost...

I formatted the code with clang-format to ensure a consistent style with the other source files.

**Comment by prokolyvakis** [comment]
Hi @RUrlus,

Unfortunately, I am testing the code in a _rather too old machine_ and I am not even able to see the performance difference that you mentioned... I performed some minor edits and I would be grateful If you could check if they have any impact on the overall computation time. My educative guess is that the abstraction provided by the `Dip class` could create a performance drop since there are now many obsolete method calls. When I find more time, I will try to dig further into it!

**Comment by RUrlus** [comment]
@prokolyvakis It's a minor improvement, we're now at:
```
2.07 s ± 13.9 ms per loop (mean ± std. dev. of 100 runs, 1 loop each)
```

**Comment by RUrlus** [comment]
Hi @prokolyvakis,

I just rebased integrated the latest changes from unstable, there was an issue with passing the architecture flags for native builds. I'll run some additional tests to verify we obtain the same statistic in different cases.

Assuming that it passes, I am happy to merge in the changes a create a new release.

**Comment by RUrlus** [comment]
@prokolyvakis I just ran a simulation test with 4million different sequences, all dip values for your CPP implementation and the C implementation are, approximately, equal.

Good job and thanks again for the PR

**Comment by prokolyvakis** [comment]
@RUrlus my apologies I have just seen your comments. Thanks for the extensive testing! :) Glad that this PR has been accepted. I did not manage to look further into the performance regression, but I will try to dig into it soon!

**Pull request merged**